Fix query hang after updating LDAP roles

alezzqz opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed issue when updating of mapped LDAP roles can lead to query hangs and denial of service.
Fixes #53708

alezzqz committed:
fix query hang after updating LDAP roles

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=54548) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=54548) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=54548) it.</sub>

tavplubix commented:
Is it possible to add a test?

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #54548 from soft-stech/ClickHouse:update-roles-deadlock-rm -->
This is an automated comment for commit 64bd945fc3c5a91393eeda4c5a99ecffd5bb0117 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54548/64bd945fc3c5a91393eeda4c5a99ecffd5bb0117/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

alezzqz commented:
> Is it possible to add a test?

If we could emulate LDAP server inside integration test environment and manage its roles - yes, it is possible.
Either we could create just unit test?

alezzqz commented:
> However, I did not find any tests for LDAP. I thought we already had some integration tests for this feature, but we do not.

So did I. That's why I'm asking.
Well, will try to make unit test

Enmk commented:
There are some test regarding LDAP that are waiting to be reviewed, merged (together with a feature). https://github.com/ClickHouse/ClickHouse/pull/42537

I believe that tests can be extracted from that PR and we might start adding some test coverage for LDAP little by little.

alezzqz commented:
changes v3:
- fixed build warnings/errors

Enmk commented on src/Access/tests/gtest_ldap_access_storage.cpp:
```diff
@@ -0,0 +1,138 @@
+    access_control.setExternalAuthenticatorsConfig(*config);
+
+    auto ldap_storage = std::make_shared<LDAPAccessStorageTest>("ldap", access_control, *config, "");
+    Mock::AllowLeak(&*ldap_storage);
```
This deserves an explanation in the comments

alezzqz commented on src/Access/LDAPAccessStorage.cpp:
```diff
@@ -312,7 +312,7 @@ void LDAPAccessStorage::updateAssignedRolesNoLock(const UUID & id, const String
     };
 
-    memory_storage.update(id, update_func);
+    memory_storage.updateNoLock(id, update_func, true);
```
The problem occurs when we lock mutex of `LDAPAccessStorage` in `authenticateImpl` and then call `update` from `updateAssignedRolesNoLock` using the same mutex - since `AccessControl` contains the same instance of `LDAPAccessStorage`. There are different solutions possible, but I guess this one is simplest and reliable.
Seems there's no danger of modifying data from different thread here since this call always locked in `LDAPAccessStorage::authenticateImpl` and `updateNoLock` can be called from single thread only.
Please correct me if I'm wrong.

alezzqz replied:
Yes, seems this is the same scenario. It was the first solution for me but I decided not to use it because using recursive mutex is not the safest idea. Well anyway it's your choice. Both solutions resolve the issue.

alezzqz commented on src/Access/MemoryAccessStorage.h:
```diff
@@ -36,6 +36,8 @@ class MemoryAccessStorage : public IAccessStorage
     bool isBackupAllowed() const override { return backup_allowed; }
     void restoreFromBackup(RestorerFromBackup & restorer) override;
 
+    bool updateNoLock(const UUID & id, const UpdateFunc & update_func, bool throw_if_not_exists) TSA_NO_THREAD_SAFETY_ANALYSIS;
```
usually I do not like using friend classes because this makes code readability worse (just in my opinion)
but your explanation is clear and I agree and used `LDAPAccessStorage` as friend class instead of making `updateNoLock` public. Done.

alezzqz commented:
changes v4:
- comments added
- renamed objects as requested
- used LDAPAccessStorage as friend class to keep updateNoLock private
- rebased on latest master
